modules/fedimint-mint-client/src/issuance.rs

```diff
@@ -0,0 +1,372 @@
+            _ => panic!("Unexpected prior state"),
+        };
+        let notes_res = bsig_res.and_then(|bsigs| {
+            issuance
+                .finalize(bsigs, &mint_keys)
+                .map_err(|e| e.to_string())
+        });
+
+        match notes_res {
+            Ok(notes) => {
+                for (amount, note) in notes.iter_items() {
+                    let replaced = dbtx
+                        .module_tx()
+                        .insert_entry(
+                            &NoteKey {
+                                amount,
+                                nonce: note.note.0,
+                            },
+                            note,
+                        )
+                        .await;
+                    if let Some(note) = replaced {
+                        error!(
+                            ?note,
+                            "E-cash note was replaced in DB, this should never happen!"
+                        )
+                    }
+                }
+                MintIssuanceStateMachine {
+                    common: old_state.common,
+                    state: MintIssuanceStates::Succeeded(MintIssuanceStatesSucceeded {
+                        amount: notes.total_amount(),
+                    }),
+                }
+            }
+            Err(error) => MintIssuanceStateMachine {
+                common: old_state.common,
+                state: MintIssuanceStates::Failed(MintIssuanceStatesFailed { error }),
+            },
+        }
+    }
+}
+
+/// See [`MintIssuanceStates`]
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceStatesAborted;
+
+/// See [`MintIssuanceStates`]
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceStatesFailed {
+    error: String,
+}
+
+/// See [`MintIssuanceStates`]
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceStatesSucceeded {
+    amount: Amount,
+}
+
+/// Single [`Note`] issuance request to the mint.f
+///
+/// Keeps the data to generate [`SpendableNote`] once the
+/// mint successfully processed the transaction signing the corresponding
+/// [`BlindNonce`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize, Encodable, Decodable)]
+pub struct NoteIssuanceRequest {
+    /// Spend key from which the note nonce (corresponding public key) is
+    /// derived
+    spend_key: KeyPair,
+    /// Key to unblind the blind signature supplied by the mint for this note
+    blinding_key: BlindingKey,
+}
+
+impl NoteIssuanceRequest {
+    /// Generate a request session for a single note and returns it plus the
+    /// corresponding blinded message
+    pub(crate) fn new<C>(
+        ctx: &Secp256k1<C>,
+        secret: DerivableSecret,
+    ) -> (NoteIssuanceRequest, BlindNonce)
+    where
+        C: Signing,
+    {
+        let spend_key = secret.child_key(ChildId(0)).to_secp_key(ctx);
```

Comment:
`ChildId(0)` and `ChildId(1)` are a confusing magical constants, that would be better of as named `const`s or something like that.